apps/language_server/lib/language_server/providers/code_lens/test.ex

```diff
@@ -0,0 +1,94 @@
+defmodule ElixirLS.LanguageServer.Providers.CodeLens.Test do
+  alias ElixirLS.LanguageServer.Providers.CodeLens
+  alias ElixirLS.LanguageServer.SourceFile
+  alias ElixirSense.Core.Parser
+  alias ElixirSense.Core.Metadata
+
+  @run_test_command "elixir.test.run"
+
+  def code_lens(uri, text) do
+    buffer_file_metadata =
+      text
+      |> Parser.parse_string(true, true, 1)
+
+    file_path = SourceFile.path_from_uri(uri)
+
+    function_lenses = get_function_lenses(buffer_file_metadata, file_path)
+    module_lenses = get_module_lenses(buffer_file_metadata, file_path)
+
+    {:ok, function_lenses ++ module_lenses}
+  end
+
+  defp get_module_lenses(%Metadata{} = metadata, file_path) do
+    metadata
+    |> get_test_modules()
+    |> Enum.map(&build_test_module_code_lens(file_path, &1))
+  end
+
+  defp get_test_modules(%Metadata{lines_to_env: lines_to_env}) do
+    lines_to_env
+    |> Enum.group_by(fn {_line, env} -> env.module end)
+    |> Enum.filter(fn {_module, module_lines_to_env} -> is_test_module?(module_lines_to_env) end)
+    |> Enum.map(fn {module, [{line, _env} | _rest]} -> {module, line} end)
+  end
+
+  defp get_function_lenses(%Metadata{} = metadata, file_path) do
+    runnable_functions = [{:test, 3}, {:test, 2}, {:describe, 2}]
+
+    calls_list =
+      metadata.calls
+      |> Enum.map(fn {_k, v} -> v end)
+      |> List.flatten()
+
+    for func <- runnable_functions do
```

Comment:
extract metadata.lines_to_env |> Map.to_list to the outside of `for` expressions. Btw those fors can be combined into one
```
for func <- runnable_functions,
{line, _col} <- calls_to(calls_list, func),
is_test_module?(metadata.lines_to_env, line) do
  build_function_test_code_lens(func, file_path, line)
end
```
without the need to flatten